File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0] + [-1] * amount
        for i in range(1, amount + 1):
            val = 100000
            for coin in coins:
                if i - coin >= 0 and dp[i - coin] != -1:
                    val = min(val, 1 + dp[i - coin])
            if val != 100000: dp[i] = val
        return dp[-1]

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int dp[10001] = {};
        fill(dp + 1, dp + 10001, 10001);
        for (int i = 1; i <= amount; ++i)
            for (int x : coins)
                if (x <= i) dp[i] = min(dp[i], 1 + dp[i - x]);
        return dp[amount] == 10001 ? -1 : dp[amount];
    }
};
